### PR TITLE
 Fix: --wait flag in ipfs-cluster-ctl and WaitFor function

### PR DESCRIPTION
**Pull request by hsanjuan**
@lanzafame this was very visibly not working when I tried it. Please test a bit better next time.

### CODE REVIEW

**Comment by lanzafame** [comment]
@hsanjuan Agreed. Could you help me see where I went wrong with the following sharness tests, as they shouldn't have passed.

From `sharness/t0030-ctl-pin.sh`:
```
expecting success:
    cid=`docker exec ipfs sh -c "dd if=/dev/urandom bs=1024 count=2048 | ipfs add -q"`
    ipfs-cluster-ctl pin add --wait "$cid" | grep -q -i "PINNED" &&
    ipfs-cluster-ctl pin ls "$cid" | grep -q "$cid" &&
    ipfs-cluster-ctl status "$cid" | grep -q -i "PINNED"

expecting success:
    cid=`ipfs-cluster-ctl --enc=json pin ls | jq -r ".[] | .cid" | head -1`
    ipfs-cluster-ctl pin rm --wait "$cid" | grep -q -i "UNPINNED" &&
    !(ipfs-cluster-ctl pin ls "$cid" | grep -q "$cid") &&
    ipfs-cluster-ctl status "$cid" | grep -q -i "UNPINNED"

expecting success:
    cid=`docker exec ipfs sh -c "dd if=/dev/urandom bs=1024 count=2048 | ipfs add -q"`
    ipfs-cluster-ctl pin add --wait --wait-timeout 2s "$cid" | grep -q -i "PINNED" &&
    ipfs-cluster-ctl pin ls "$cid" | grep -q "$cid" &&
    ipfs-cluster-ctl status "$cid" | grep -q -i "PINNED"

expecting success:
    cid=`ipfs-cluster-ctl --enc=json pin ls | jq -r ".[] | .cid" | head -1`
    ipfs-cluster-ctl pin rm --wait --wait-timeout 2s "$cid" | grep -q -i "UNPINNED" &&
    !(ipfs-cluster-ctl pin ls "$cid" | grep -q "$cid") &&
    ipfs-cluster-ctl status "$cid" | grep -q -i "UNPINNED"
```

Also, `api/rest/client/methods_test.go` is failing to build because a reference to `test.SlowCid` which doesn't exist?

File: ipfs-cluster-ctl/main.go [hunk 1]
```diff
@@ -712,3 +674,54 @@ func parseCredentials(userInput string) (string, string) {
...
+		status, cerr = waitFor(ci, target, timeout)
+		checkErr("waiting for pin status", cerr)
+	}
+
+	if noStatus {
+		return
+	}
+
+	if status.Cid == nil { // no status from "wait"
+		time.Sleep(1000 * time.Millisecond)
```
Comment (lanzafame) [inline ipfs-cluster-ctl/main.go]: `s/1000*time.Millisecond/1*time.Second`

File: ipfs-cluster-ctl/main.go [hunk 2]
```diff
@@ -398,34 +380,14 @@ although unpinning operations in the cluster may take longer or fail.
...
-							}
-
-							status, cerr := globalClient.WaitFor(ctx, fp)
-							formatResponse(c, status, cerr)
-						}
-
+						waitAndFormatStatusResponse(
+							c,
+							c.Bool("wait"),
+							c.Bool("no-status"),
```
Comment (lanzafame) [inline ipfs-cluster-ctl/main.go]: It seems weird that `waitAndFormatStatusResponse` would handle the `--no-status` flag. Maybe `handleResponseFormatFlags` would be a better name and then retrieve `wait` and `no-status` from `*cli.Context`. Or the other way, could be to handle `no-status` outside of `waitAndFormatStatusResponse`. Not sure which would be better? @hsanjuan

File: api/rest/client/methods_test.go [hunk 3]
```diff
@@ -308,3 +317,115 @@ func TestGetConnectGraph(t *testing.T) {
...
+	rpcC := rpc.NewClientWithServer(nil, "wait", rpcS)
+	err := rpcS.RegisterName("Cluster", &waitService{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tapi.SetClient(rpcC)
+
+	testF := func(t *testing.T, c *Client) {
+		ci, _ := cid.Decode(test.SlowCid)
```
Comment (lanzafame) [inline api/rest/client/methods_test.go]: `test.SlowCid` doesn't exist. Related question, where do the `test.TestCid1`, etc, come from when running the tests?

File: api/rest/client/methods.go [hunk 4]
```diff
@@ -251,12 +252,10 @@ func (sf *statusFilter) filter(ctx context.Context, fp StatusFilterParams) {
 				return
 			}
 
+			sf.Out <- gblPinInfo
```
Comment (lanzafame) [inline api/rest/client/methods.go]: 👍

File: api/rest/client/methods.go [hunk 5]
```diff
@@ -191,18 +191,19 @@ func (c *Client) WaitFor(ctx context.Context, fp StatusFilterParams) (api.Global
...
 		case err := <-sf.Err:
 			return api.GlobalPinInfo{}, err
-		case <-sf.Done:
-			// drain status channel and return last status on channel
-			for {
-				if status, ok := <-sf.Out; !ok {
-					return status, nil
-				}
+		case st, ok := <-sf.Out:
+			if ok {
```
Comment (lanzafame) [inline api/rest/client/methods.go]: Could be simplified to:
```go
if !ok {
    return status, nil
}
status = st
```

File: ipfs-cluster-ctl/main.go (hunk 2 above)
Comment (hsanjuan) [reply ipfs-cluster-ctl/main.go]: did the first

**Comment by hsanjuan** [comment]
> @hsanjuan Agreed. Could you help me see where I went wrong with the following sharness tests, as they shouldn't have passed.

The first case in a switch was `case !c.Bool("no-status")`. Since that flag is set to false by default, it waited 1 second and print a status (in which everything was pinned already because it's fast enough). So the output would look like you had been waiting.

While those tests verify that the flags are actually valid and usable and that we can call the command with them, they don't fully verify the behavior. I haven't come up with a straightforward way to fully verify `--wait` with sharness so I left those as they are. They still will catch some errors.

File: ipfs-cluster-ctl/main.go [hunk 6]
```diff
@@ -712,3 +670,53 @@ func parseCredentials(userInput string) (string, string) {
 		return "", ""
 	}
 }
+
+func handlePinResponseFormatFlags(
+	c *cli.Context,
+	ci *cid.Cid,
+	target api.TrackerStatus,
+	timeout time.Duration,
```
Comment (lanzafame) [inline ipfs-cluster-ctl/main.go]: We could access the timeout value via `cli.Context`?

File: ipfs-cluster-ctl/main.go [hunk 7]
```diff
@@ -712,3 +670,53 @@ func parseCredentials(userInput string) (string, string) {
...
+	if c.Bool("wait") {
+		status, cerr = waitFor(ci, target, timeout)
+		checkErr("waiting for pin status", cerr)
+	}
+
+	if c.Bool("no-status") {
+		return
+	}
+
+	if status.Cid == nil { // no status from "wait"
```
Comment (lanzafame) [inline ipfs-cluster-ctl/main.go]: Could this go before the `no-status` check as it is related to the `wait` flag?

Comment (lanzafame) [reply ipfs-cluster-ctl/main.go]: Does having `--wait` and `--no-status` set at the same time make sense though?

Comment (lanzafame) [reply ipfs-cluster-ctl/main.go]: sorry nvm, I totally not understanding that correctly

**Pull request merged**